File: desktop_client/assets.py
from __future__ import annotations

import hashlib
import json
import mimetypes
import os
import platform
import shutil
import stat
import subprocess
import sys
import tempfile
import threading
import zipfile
from dataclasses import asdict, dataclass
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path, PurePosixPath
from typing import BinaryIO, cast
from urllib.error import HTTPError, URLError
from urllib.parse import unquote, urlsplit
from urllib.request import Request, urlopen


ASSET_VERSION_PATH = "/api/editor_version"
ASSET_BUNDLE_PATH = "/api/editor_assets_bundle"
ASSET_CONNECT_TIMEOUT_SECONDS = 3
ASSET_DOWNLOAD_TIMEOUT_SECONDS = 120

DATA_HOME = Path(
    os.environ.get("XDG_DATA_HOME", Path.home() / ".local" / "share")
)
CACHE_HOME = Path(os.environ.get("XDG_CACHE_HOME", Path.home() / ".cache"))
ASSET_ROOT = DATA_HOME / "te2" / "desktop_assets"
BUILD_CACHE_ROOT = CACHE_HOME / "te2" / "desktop_shell" / "web_extensions"
WEB_EXTENSION_SOURCE = (
    Path(__file__).parent / "web_process_extension" / "asset_redirect.c"
)
ASSET_INVENTORY_PATH = Path(__file__).parent / "desktop_asset_inventory.json"

# WebKitGTK and Electrobun consume this same immutable-static boundary. Keep the
# routing contract in data so the two desktop shells cannot silently drift.
with ASSET_INVENTORY_PATH.open(encoding="utf-8") as _inventory_file:
    _asset_inventory = json.load(_inventory_file)

LOCAL_PREFIXES = tuple(_asset_inventory["localPrefixes"])
LOCAL_FILES = tuple(_asset_inventory["localFiles"])
LOCAL_MAPPINGS = tuple(
    (mapping["kind"], mapping["source"], mapping["destination"])
    for mapping in _asset_inventory["localMappings"]
)
REQUIRED_DESKTOP_ASSET_FILES = tuple(_asset_inventory["requiredFiles"])


def compare_asset_versions(left: str, right: str) -> int:
    def parts(value: str) -> list[int]:
        result: list[int] = []
        for part in value.split("."):
            try:
                result.append(int(part))
            except ValueError:
                result.append(0)
        return result

    left_parts = parts(left)
    right_parts = parts(right)
    length = max(len(left_parts), len(right_parts))
    left_parts.extend([0] * (length - len(left_parts)))
    right_parts.extend([0] * (length - len(right_parts)))
    return (left_parts > right_parts) - (left_parts < right_parts)


def map_local_asset_path(request_path: str) -> str | None:
    if request_path not in LOCAL_FILES and not any(
        request_path.startswith(prefix) for prefix in LOCAL_PREFIXES
    ):
        return None

    for kind, source, destination in LOCAL_MAPPINGS:
        if kind == "exact" and request_path == source:
            return destination
        if kind == "prefix" and request_path.startswith(source):
            return destination + request_path[len(source) :]
    return request_path


def _normalize_framework_base_url(base_url: str) -> str:
    parsed = urlsplit(base_url)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise ValueError("Framework URL must be an HTTP or HTTPS origin")
    return f"{parsed.scheme}://{parsed.netloc}"


def _safe_zip_path(name: str) -> Path | None:
    candidate = PurePosixPath(name)
    if candidate.is_absolute() or ".." in candidate.parts:
        return None
    parts = tuple(part for part in candidate.parts if part not in {"", "."})
    if not parts or parts[0] == "android-shell":
        return None
    return Path(*parts)


@dataclass(frozen=True)
class AssetUpdateResult:
    ok: bool
    updated: bool
    local_version: str | None
    server_version: str | None = None
    error: str | None = None

    def to_dict(self) -> dict[str, object]:
        return asdict(self)


@dataclass(frozen=True)
class WebExtensionBuild:
    available: bool
    directory: Path | None = None
    library: Path | None = None
    error: str | None = None


class DesktopAssetManager:
    def __init__(self, asset_root: Path = ASSET_ROOT) -> None:
        self.asset_root = asset_root
        self._lock = threading.RLock()

    @property
    def version_file(self) -> Path:
        return self.asset_root / "version.txt"

    def local_version(self) -> str | None:
        try:
            value = self.version_file.read_text(encoding="utf-8").strip()
        except OSError:
            return None
        return value or None

    def missing_required_asset(self, root: Path | None = None) -> str | None:
        target = root or self.asset_root
        return next(
            (
                relative
                for relative in REQUIRED_DESKTOP_ASSET_FILES
                if not (target / relative).is_file()
            ),
            None,
        )

    def has_valid_assets(self) -> bool:
        return self.missing_required_asset() is None

    def status(self) -> dict[str, object]:
        missing = self.missing_required_asset()
        return {
            "assetRoot": str(self.asset_root),
            "localVersion": self.local_version(),
            "valid": missing is None,
            "missingRequiredAsset": missing,
        }

    def fetch_server_version(self, base_url: str) -> str:
        request = Request(
            _normalize_framework_base_url(base_url) + ASSET_VERSION_PATH,
            headers={"Accept": "text/plain"},
        )
        try:
            with urlopen(
                request,
                timeout=ASSET_CONNECT_TIMEOUT_SECONDS,
            ) as response:
                version = response.read().decode("utf-8", "replace").strip()
        except HTTPError as error:
            raise RuntimeError(
                f"Asset version check returned HTTP {error.code}"
            ) from error
        except (URLError, OSError) as error:
            reason = getattr(error, "reason", error)
            raise RuntimeError(f"Asset version check failed: {reason}") from error
        if not version:
            raise RuntimeError("Asset version check returned an empty version")
        return version

    def update_from_server(
        self,
        base_url: str,
        *,
        force: bool = False,
    ) -> AssetUpdateResult:
        with self._lock:
            local_version = self.local_version()
            try:
                server_version = self.fetch_server_version(base_url)
            except RuntimeError as error:
                return AssetUpdateResult(
                    ok=False,
                    updated=False,
                    local_version=local_version,
                    error=str(error),
                )

            if local_version is not None:
                comparison = compare_asset_versions(server_version, local_version)
                if comparison < 0:
                    return AssetUpdateResult(
                        ok=False,
                        updated=False,
                        local_version=local_version,
                        server_version=server_version,
                        error=(
                            "Refusing asset downgrade "
                            f"from {local_version} to {server_version}"
                        ),
                    )
                if comparison == 0 and self.has_valid_assets() and not force:
                    return AssetUpdateResult(
                        ok=True,
                        updated=False,
                        local_version=local_version,
                        server_version=server_version,
                    )

            try:
                installed_version = self._download_and_install(
                    base_url,
                    local_version=local_version,
                )
            except (OSError, RuntimeError, zipfile.BadZipFile) as error:
                return AssetUpdateResult(
                    ok=False,
                    updated=False,
                    local_version=self.local_version(),
                    server_version=server_version,
                    error=str(error),
                )

            return AssetUpdateResult(
                ok=True,
                updated=True,
                local_version=installed_version,
                server_version=server_version,
            )

    def _download_and_install(
        self,
        base_url: str,
        *,
        local_version: str | None,
    ) -> str:
        parent = self.asset_root.parent
        parent.mkdir(parents=True, exist_ok=True)
        staging = parent / f".{self.asset_root.name}.staging"
        backup = parent / f".{self.asset_root.name}.backup"
        for path in (staging, backup):
            if path.exists():
                shutil.rmtree(path)

        bundle_path: Path | None = None
        try:
            with tempfile.NamedTemporaryFile(
                prefix=".desktop-assets-",
                suffix=".zip",
                dir=parent,
                delete=False,
            ) as bundle:
                bundle_path = Path(bundle.name)
                self._download_bundle(
                    base_url,
                    cast(BinaryIO, cast(object, bundle)),
                )

            staging.mkdir(parents=True)
            with zipfile.ZipFile(bundle_path) as archive:
                for info in archive.infolist():
                    mode = info.external_attr >> 16
                    if stat.S_ISLNK(mode):
                        raise RuntimeError(
                            f"Asset bundle contains a symbolic link: {info.filename}"
                        )
                    relative = _safe_zip_path(info.filename)
                    if relative is None:
                        continue
                    target = staging / relative
                    if info.is_dir():
                        target.mkdir(parents=True, exist_ok=True)
                        continue
                    target.parent.mkdir(parents=True, exist_ok=True)
                    with archive.open(info) as source, target.open("wb") as output:
                        shutil.copyfileobj(source, output)

            staged_version_file = staging / "version.txt"
            try:
                staged_version = staged_version_file.read_text(
                    encoding="utf-8"
                ).strip()
            except OSError as error:
                raise RuntimeError(
                    "Asset bundle is missing version.txt"
                ) from error
            if not staged_version:
                raise RuntimeError("Asset bundle contains an empty version.txt")

            missing = self.missing_required_asset(staging)
            if missing is not None:
                raise RuntimeError(
                    f"Asset bundle is missing required file: {missing}"
                )
            if (
                local_version is not None
                and compare_asset_versions(staged_version, local_version) < 0
            ):
                raise RuntimeError(
                    "Refusing staged asset downgrade "
                    f"from {local_version} to {staged_version}"
                )

            if self.asset_root.exists():
                self.asset_root.rename(backup)
            try:
                staging.rename(self.asset_root)
            except OSError:
                if backup.exists() and not self.asset_root.exists():
                    backup.rename(self.asset_root)
                raise
            if backup.exists():
                shutil.rmtree(backup, ignore_errors=True)
            return staged_version
        finally:
            if bundle_path is not None:
                bundle_path.unlink(missing_ok=True)
            if staging.exists():
                shutil.rmtree(staging)

    def _download_bundle(self, base_url: str, output: BinaryIO) -> None:
        request = Request(
            _normalize_framework_base_url(base_url) + ASSET_BUNDLE_PATH,
            headers={"Accept": "application/zip"},
        )
        try:
            with urlopen(
                request,
                timeout=ASSET_DOWNLOAD_TIMEOUT_SECONDS,
            ) as response:
                shutil.copyfileobj(response, output)
        except HTTPError as error:
            raise RuntimeError(
                f"Asset download returned HTTP {error.code}"
            ) from error
        except (URLError, OSError) as error:
            reason = getattr(error, "reason", error)
            raise RuntimeError(f"Asset download failed: {reason}") from error


class _AssetRequestHandler(BaseHTTPRequestHandler):
    server_version = "TE2DesktopAssets/1"

    def do_GET(self) -> None:
        self._serve(send_body=True)

    def do_HEAD(self) -> None:
        self._serve(send_body=False)

    def _serve(self, *, send_body: bool) -> None:
        server = self.server
        if not isinstance(server, _DesktopAssetHTTPServer):
            self.send_error(500)
            return
        raw_path = unquote(urlsplit(self.path).path)
        relative = _safe_zip_path(raw_path.lstrip("/"))
        if relative is None:
            self.send_error(403)
            return
        requested = (server.asset_root / relative).resolve()
        root = server.asset_root.resolve()
        try:
            requested.relative_to(root)
        except ValueError:
            self.send_error(403)
            return
        if not requested.is_file():
            self.send_error(404)
            return

        mime_type = mimetypes.guess_type(requested.name)[0]
        size = requested.stat().st_size
        self.send_response(200)
        self.send_header("Content-Type", mime_type or "application/octet-stream")
        self.send_header("Content-Length", str(size))
        self.send_header("Cache-Control", "max-age=31536000, immutable")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        if send_body:
            with requested.open("rb") as source:
                shutil.copyfileobj(source, self.wfile)

    def log_message(self, format: str, *args: object) -> None:
        del format, args
        return


class _DesktopAssetHTTPServer(ThreadingHTTPServer):
    daemon_threads = True
    allow_reuse_address = True

    def __init__(self, asset_root: Path) -> None:
        self.asset_root = asset_root
        super().__init__(("127.0.0.1", 0), _AssetRequestHandler)

    def handle_error(self, request, client_address) -> None:
        error = sys.exc_info()[1]
        if isinstance(error, (BrokenPipeError, ConnectionResetError)):
            return
        super().handle_error(request, client_address)


class DesktopAssetServer:
    def __init__(self, asset_root: Path = ASSET_ROOT) -> None:
        self.asset_root = asset_root
        self._server: _DesktopAssetHTTPServer | None = None
        self._thread: threading.Thread | None = None

    @property
    def base_url(self) -> str | None:
        if self._server is None:
            return None
        host, port = self._server.server_address[:2]
        return f"http://{host}:{port}"

    def start(self) -> str:
        if self._server is not None:
            return self.base_url or ""
        self.asset_root.parent.mkdir(parents=True, exist_ok=True)
        server = _DesktopAssetHTTPServer(self.asset_root)
        thread = threading.Thread(
            target=server.serve_forever,
            name="te2-desktop-assets",
            daemon=True,
        )
        thread.start()
        self._server = server
        self._thread = thread
        return self.base_url or ""

    def stop(self) -> None:
        server = self._server
        thread = self._thread
        self._server = None
        self._thread = None
        if server is None:
            return
        server.shutdown()
        server.server_close()
        if thread is not None:
            thread.join(timeout=2)


def _prune_old_web_extension_builds(active_directory: Path) -> None:
    try:
        candidates = tuple(BUILD_CACHE_ROOT.iterdir())
    except OSError:
        return
    for candidate in candidates:
        if candidate != active_directory and candidate.is_dir():
            shutil.rmtree(candidate, ignore_errors=True)


def ensure_web_extension() -> WebExtensionBuild:
    if not WEB_EXTENSION_SOURCE.is_file():
        return WebExtensionBuild(
            available=False,
            error=f"Missing WebKit extension source: {WEB_EXTENSION_SOURCE}",
        )
    try:
        version = subprocess.run(
            ["pkg-config", "--modversion", "webkitgtk-web-process-extension-6.0"],
            check=True,
            capture_output=True,
            text=True,
        ).stdout.strip()
        flags = subprocess.run(
            [
                "pkg-config",
                "--cflags",
                "--libs",
                "webkitgtk-web-process-extension-6.0",
            ],
            check=True,
            capture_output=True,
            text=True,
        ).stdout.split()
    except (FileNotFoundError, subprocess.CalledProcessError) as error:
        detail = getattr(error, "stderr", None) or str(error)
        return WebExtensionBuild(
            available=False,
            error=(
                "WebKit asset interception needs a C compiler and the "
                "WebKitGTK 6 development package: "
                f"{str(detail).strip()}"
            ),
        )

    digest = hashlib.sha256()
    digest.update(WEB_EXTENSION_SOURCE.read_bytes())
    digest.update(version.encode())
    digest.update(platform.machine().encode())
    fingerprint = digest.hexdigest()[:16]
    directory = BUILD_CACHE_ROOT / fingerprint
    library = directory / "te2_desktop_asset_redirect.so"
    metadata = directory / "build.json"
    if library.is_file() and metadata.is_file():
        _prune_old_web_extension_builds(directory)
        return WebExtensionBuild(True, directory, library)

    directory.mkdir(parents=True, exist_ok=True)
    temporary = directory / ".te2_desktop_asset_redirect.so.tmp"
    command = [
        os.environ.get("CC", "cc"),
        "-std=c11",
        "-shared",
        "-fPIC",
        "-O2",
        "-Wall",
        "-Wextra",
        "-o",
        str(temporary),
        str(WEB_EXTENSION_SOURCE),
        *flags,
    ]
    try:
        completed = subprocess.run(
            command,
            check=True,
            capture_output=True,
            text=True,
        )
        temporary.replace(library)
        metadata.write_text(
            json.dumps(
                {
                    "source": str(WEB_EXTENSION_SOURCE),
                    "webkitVersion": version,
                    "machine": platform.machine(),
                    "compilerOutput": completed.stderr.strip(),
                },
                indent=2,
            ),
            encoding="utf-8",
        )
    except (OSError, subprocess.CalledProcessError) as error:
        temporary.unlink(missing_ok=True)
        detail = getattr(error, "stderr", None) or str(error)
        return WebExtensionBuild(
            available=False,
            error=f"Failed to build WebKit asset interceptor: {str(detail).strip()}",
        )
    _prune_old_web_extension_builds(directory)
    return WebExtensionBuild(True, directory, library)

File: desktop_client/electrobun_spike/src/bun/assets.ts
import { createWriteStream } from "node:fs";
import {
  access,
  mkdir,
  readFile,
  rename,
  rm,
  stat,
} from "node:fs/promises";
import { homedir } from "node:os";
import { dirname, extname, join, normalize, relative, resolve, sep } from "node:path";
import { pipeline } from "node:stream/promises";
import { Readable } from "node:stream";

import yauzl, { type Entry, type ZipFile } from "yauzl";

import inventoryData from "../../../desktop_asset_inventory.json";

const ASSET_VERSION_PATH = "/api/editor_version";
const ASSET_BUNDLE_PATH = "/api/editor_assets_bundle";
const CONNECT_TIMEOUT_MS = 3_000;
const DOWNLOAD_TIMEOUT_MS = 120_000;

type MappingKind = "exact" | "prefix";

export type AssetMapping = {
  kind: MappingKind;
  source: string;
  destination: string;
};

export type AssetInventory = {
  localPrefixes: string[];
  localFiles: string[];
  localMappings: AssetMapping[];
  requiredFiles: string[];
};

export const ASSET_INVENTORY = inventoryData as AssetInventory;

export type AssetStatus = {
  assetRoot: string;
  localVersion: string | null;
  valid: boolean;
  missingRequiredAsset: string | null;
  serverBaseUrl: string | null;
  interceptorAvailable: boolean;
  interceptorError: string | null;
};

export type AssetUpdateResult = AssetStatus & {
  ok: boolean;
  updated: boolean;
  serverVersion: string | null;
  error: string | null;
};

function dataHome(environment = process.env): string {
  return environment.XDG_DATA_HOME?.trim() || join(homedir(), ".local", "share");
}

export function desktopAssetRoot(environment = process.env): string {
  return join(dataHome(environment), "te2", "desktop_assets");
}

export function compareAssetVersions(left: string, right: string): number {
  const parts = (value: string) =>
    value.split(".").map((part) => {
      const parsed = Number.parseInt(part, 10);
      return Number.isFinite(parsed) ? parsed : 0;
    });
  const leftParts = parts(left);
  const rightParts = parts(right);
  const length = Math.max(leftParts.length, rightParts.length);
  for (let index = 0; index < length; index += 1) {
    const difference = (leftParts[index] || 0) - (rightParts[index] || 0);
    if (difference !== 0) return difference > 0 ? 1 : -1;
  }
  return 0;
}

export function mapLocalAssetPath(requestPath: string): string | null {
  if (
    !ASSET_INVENTORY.localFiles.includes(requestPath) &&
    !ASSET_INVENTORY.localPrefixes.some((prefix) => requestPath.startsWith(prefix))
  ) {
    return null;
  }
  for (const mapping of ASSET_INVENTORY.localMappings) {
    if (mapping.kind === "exact" && requestPath === mapping.source) {
      return mapping.destination;
    }
    if (mapping.kind === "prefix" && requestPath.startsWith(mapping.source)) {
      return `${mapping.destination}${requestPath.slice(mapping.source.length)}`;
    }
  }
  return requestPath;
}

export function nativeAssetRedirectRules(serverBaseUrl: string, assetRoot?: string): string {
  const mappings = new Map<string, AssetMapping>();
  for (const path of ASSET_INVENTORY.localFiles) {
    mappings.set(`exact\t${path}`, {
      kind: "exact",
      source: path,
      destination: mapLocalAssetPath(path) || path,
    });
  }
  for (const path of ASSET_INVENTORY.localPrefixes) {
    const configured = ASSET_INVENTORY.localMappings.find(
      (mapping) => mapping.kind === "prefix" && path.startsWith(mapping.source),
    );
    mappings.set(`prefix\t${path}`, {
      kind: "prefix",
      source: path,
      destination: configured
        ? `${configured.destination}${path.slice(configured.source.length)}`
        : path,
    });
  }
  return [
    `base\t${serverBaseUrl}`,
    ...(assetRoot ? [`root\t${resolve(assetRoot)}`] : []),
    ...Array.from(mappings.values(), (mapping) =>
      `${mapping.kind}\t${mapping.source}\t${mapping.destination}`,
    ),
  ].join("\n");
}

function safeArchivePath(name: string): string | null {
  const posix = name.replaceAll("\\", "/");
  if (posix.startsWith("/") || posix.split("/").includes("..")) return null;
  const parts = posix.split("/").filter((part) => part && part !== ".");
  if (!parts.length || parts[0] === "android-shell") return null;
  return parts.join(sep);
}

function openZip(path: string): Promise<ZipFile> {
  return new Promise((resolvePromise, reject) => {
    yauzl.open(path, { lazyEntries: true }, (error, zipFile) => {
      if (error || !zipFile) reject(error || new Error("Failed to open asset bundle"));
      else resolvePromise(zipFile);
    });
  });
}

function openEntryStream(zipFile: ZipFile, entry: Entry): Promise<NodeJS.ReadableStream> {
  return new Promise((resolvePromise, reject) => {
    zipFile.openReadStream(entry, (error, stream) => {
      if (error || !stream) reject(error || new Error(`Failed to read ${entry.fileName}`));
      else resolvePromise(stream);
    });
  });
}

async function extractZip(bundlePath: string, destination: string): Promise<void> {
  const zipFile = await openZip(bundlePath);
  await new Promise<void>((resolvePromise, reject) => {
    let settled = false;
    const fail = (error: unknown) => {
      if (settled) return;
      settled = true;
      zipFile.close();
      reject(error);
    };
    zipFile.on("error", fail);
    zipFile.on("end", () => {
      if (settled) return;
      settled = true;
      resolvePromise();
    });
    zipFile.on("entry", (entry) => {
      void (async () => {
        const mode = (entry.externalFileAttributes >>> 16) & 0xffff;
        if ((mode & 0o170000) === 0o120000) {
          throw new Error(`Asset bundle contains a symbolic link: ${entry.fileName}`);
        }
        const safe = safeArchivePath(entry.fileName);
        if (!safe) {
          zipFile.readEntry();
          return;
        }
        const target = join(destination, safe);
        const targetRelative = relative(resolve(destination), resolve(target));
        if (targetRelative.startsWith("..") || targetRelative === "") {
          throw new Error(`Asset bundle path escapes staging: ${entry.fileName}`);
        }
        if (entry.fileName.endsWith("/")) {
          await mkdir(target, { recursive: true });
        } else {
          await mkdir(dirname(target), { recursive: true });
          const stream = await openEntryStream(zipFile, entry);
          await pipeline(stream, createWriteStream(target));
        }
        zipFile.readEntry();
      })().catch(fail);
    });
    zipFile.readEntry();
  });
}

const MIME_TYPES: Record<string, string> = {
  ".css": "text/css; charset=utf-8",
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".mjs": "text/javascript; charset=utf-8",
  ".png": "image/png",
  ".svg": "image/svg+xml",
  ".ttf": "font/ttf",
  ".wasm": "application/wasm",
  ".woff": "font/woff",
  ".woff2": "font/woff2",
};

export class DesktopAssetManager {
  readonly assetRoot: string;
  private server: ReturnType<typeof Bun.serve> | null = null;
  private updatePromise: Promise<AssetUpdateResult> | null = null;

  constructor(assetRoot = desktopAssetRoot()) {
    this.assetRoot = assetRoot;
  }

  get serverBaseUrl(): string | null {
    return this.server ? `http://${this.server.hostname}:${this.server.port}` : null;
  }

  async localVersion(root = this.assetRoot): Promise<string | null> {
    try {
      const value = (await readFile(join(root, "version.txt"), "utf8")).trim();
      return value || null;
    } catch {
      return null;
    }
  }

  async missingRequiredAsset(root = this.assetRoot): Promise<string | null> {
    for (const file of ASSET_INVENTORY.requiredFiles) {
      try {
        await access(join(root, file));
      } catch {
        return file;
      }
    }
    return null;
  }

  async status(): Promise<AssetStatus> {
    const missing = await this.missingRequiredAsset();
    return {
      assetRoot: this.assetRoot,
      localVersion: await this.localVersion(),
      valid: missing === null,
      missingRequiredAsset: missing,
      serverBaseUrl: this.serverBaseUrl,
      interceptorAvailable: true,
      interceptorError: null,
    };
  }

  async startServer(): Promise<string> {
    if (this.server) return this.serverBaseUrl!;
    await mkdir(this.assetRoot, { recursive: true });
    const root = resolve(this.assetRoot);
    this.server = Bun.serve({
      hostname: "127.0.0.1",
      port: 0,
      fetch: async (request) => {
        const url = new URL(request.url);
        if (process.env.TE2_DESKTOP_ASSET_TRACE === "1") {
          console.log(`[te2-desktop-asset] ${request.method} ${url.pathname}`);
        }
        let decoded: string;
        try {
          decoded = decodeURIComponent(url.pathname).replace(/^\/+/, "");
        } catch {
          return new Response("Bad request", { status: 400 });
        }
        const target = resolve(root, normalize(decoded));
        const targetRelative = relative(root, target);
        if (!decoded || targetRelative.startsWith("..") || targetRelative === "") {
          return new Response("Forbidden", { status: 403 });
        }
        try {
          if (!(await stat(target)).isFile()) return new Response("Not found", { status: 404 });
        } catch {
          return new Response("Not found", { status: 404 });
        }
        const headers = {
          "Access-Control-Allow-Origin": "*",
          "Cache-Control": "max-age=31536000, immutable",
          "Content-Type": MIME_TYPES[extname(target).toLowerCase()] || "application/octet-stream",
        };
        if (request.method === "HEAD") return new Response(null, { headers });
        return new Response(Bun.file(target), { headers });
      },
    });
    return this.serverBaseUrl!;
  }

  stopServer(): void {
    this.server?.stop(true);
    this.server = null;
  }

  updateFromServer(baseUrl: string, force = false): Promise<AssetUpdateResult> {
    if (this.updatePromise) return this.updatePromise;
    this.updatePromise = this.performUpdate(baseUrl, force).finally(() => {
      this.updatePromise = null;
    });
    return this.updatePromise;
  }

  private async performUpdate(baseUrl: string, force: boolean): Promise<AssetUpdateResult> {
    const localVersion = await this.localVersion();
    let serverVersion: string | null = null;
    try {
      const response = await fetch(new URL(ASSET_VERSION_PATH, `${baseUrl}/`), {
        headers: { Accept: "text/plain" },
        signal: AbortSignal.timeout(CONNECT_TIMEOUT_MS),
      });
      if (!response.ok) throw new Error(`Asset version check returned HTTP ${response.status}`);
      serverVersion = (await response.text()).trim();
      if (!serverVersion) throw new Error("Asset version check returned an empty version");
      if (localVersion) {
        const comparison = compareAssetVersions(serverVersion, localVersion);
        if (comparison < 0) {
          throw new Error(`Refusing asset downgrade from ${localVersion} to ${serverVersion}`);
        }
        if (comparison === 0 && (await this.missingRequiredAsset()) === null && !force) {
          return { ...(await this.status()), ok: true, updated: false, serverVersion, error: null };
        }
      }
      await this.downloadAndInstall(baseUrl, localVersion);
      return { ...(await this.status()), ok: true, updated: true, serverVersion, error: null };
    } catch (error) {
      return {
        ...(await this.status()),
        ok: false,
        updated: false,
        serverVersion,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  private async downloadAndInstall(baseUrl: string, localVersion: string | null): Promise<void> {
    const parent = dirname(this.assetRoot);
    const staging = join(parent, `.${this.assetRoot.split(sep).at(-1)}.staging`);
    const backup = join(parent, `.${this.assetRoot.split(sep).at(-1)}.backup`);
    const bundle = join(parent, `.desktop-assets-${process.pid}-${Date.now()}.zip`);
    await mkdir(parent, { recursive: true });
    await Promise.all([rm(staging, { recursive: true, force: true }), rm(backup, { recursive: true, force: true })]);
    try {
      const response = await fetch(new URL(ASSET_BUNDLE_PATH, `${baseUrl}/`), {
        headers: { Accept: "application/zip" },
        signal: AbortSignal.timeout(DOWNLOAD_TIMEOUT_MS),
      });
      if (!response.ok || !response.body) {
        throw new Error(`Asset download returned HTTP ${response.status}`);
      }
      await pipeline(Readable.fromWeb(response.body as never), createWriteStream(bundle));
      await mkdir(staging, { recursive: true });
      await extractZip(bundle, staging);
      const stagedVersion = await this.localVersion(staging);
      if (!stagedVersion) throw new Error("Asset bundle is missing version.txt");
      const missing = await this.missingRequiredAsset(staging);
      if (missing) throw new Error(`Asset bundle is missing required file: ${missing}`);
      if (localVersion && compareAssetVersions(stagedVersion, localVersion) < 0) {
        throw new Error(`Refusing staged asset downgrade from ${localVersion} to ${stagedVersion}`);
      }
      try {
        await rename(this.assetRoot, backup);
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
      }
      try {
        await rename(staging, this.assetRoot);
      } catch (error) {
        try {
          await rename(backup, this.assetRoot);
        } catch {
          // Preserve the original install error when no backup existed.
        }
        throw error;
      }
      await rm(backup, { recursive: true, force: true });
    } finally {
      await Promise.all([
        rm(bundle, { force: true }),
        rm(staging, { recursive: true, force: true }),
      ]);
    }
  }
}
